无法转换一个简单的 UNet

<!-- 请您简介清晰的描述您遇到的问题 -->
## 环境
1.系统环境：ubuntu 18.04
2.MegEngine版本：MegEngine 1.1
3.python版本：3.6.9

## 复现步骤
见代码

## 请提供关键的代码片段便于追查问题

模型见
https://github.com/megvii-research/PMRID/blob/main/models/net_mge.py

先 dump 一个模型出来
```python
from megengine import jit
net = Network()
net.eval()
img = mge.tensor(np.random.randn(1, 4, 64, 64).astype(np.float32))

@jit.trace(capture_as_const=True)
def f(x):
    out = net(x)
    return out

f(img)
f.dump(
    "net.mge",
    arg_names=['img'],
    optimize_for_inference=False,
    output_names=['pred'],
)
```

## 请提供完整的日志及报错信息

在将 #4  自己修掉的情况下运行 megconverter:
```
python3 -m mgeconvert.utils.convert_onnx  -i net.mge -o net.onnx
```

```
Traceback (most recent call last):
  File "/usr/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/wangyuzhi/.local/lib/python3.6/site-packages/mgeconvert/utils/convert_onnx.py", line 29, in <module>
    main()
  File "/home/wangyuzhi/.local/lib/python3.6/site-packages/mgeconvert/utils/convert_onnx.py", line 25, in main
    convert_to_onnx(args.input, args.output, graph_name=args.graph, opset=args.opset)
  File "/home/wangyuzhi/.local/lib/python3.6/site-packages/mgeconvert/onnx_converter/onnx_converter.py", line 120, in convert_to_onnx
    model = converter.convert()
  File "/home/wangyuzhi/.local/lib/python3.6/site-packages/mgeconvert/onnx_converter/onnx_converter.py", line 70, in convert
    unsupported_oprs
AssertionError: Operators {<class 'mgeconvert.mge_context.mge_op.ConvolutionBackwardDataOpr'>} are not supported yet
```

我不了解 `ConvolutionBackwardDataOpr` 这个 opr 的作用，凭我朴素的理解，我只需要得到这个网络的前向部分，跟 backward 应该没关系……

--- a/README.md
+++ b/README.md
@@ -17,7 +17,7 @@ MgeConvert前端的部分都在`mge_context`目录下, 可以直接将MegEngine 
 |broadcast| ✓ | ✓ | ✓ |
 |concat| ✓ | ✓ | ✓ |
 |conv2d| ✓ | ✓ | ✓ |
-|convtranspose2d| × | × | ✓ |
+|convtranspose2d| ✓ | ✓ | ✓ |
 |div(true_div)| ✓ | ✓ | ✓ |
 |exp| ✓ | ✓ | ✓ |
 |log| ✓ | ✓ | ✓ |

--- a/mgeconvert/caffe_converter/caffe_op.py
+++ b/mgeconvert/caffe_converter/caffe_op.py
@@ -15,6 +15,7 @@ from ..mge_context import (
     AxisAddRemoveOpr,
     BatchNormForwardOpr,
     ConcatOpr,
+    ConvolutionBackwardDataOpr,
     ConvolutionForwardOpr,
     DimshuffleOpr,
     ElemwiseOpr,
@@ -539,14 +540,17 @@ def _eltwise(opr: ElemwiseOpr, context):
         )
 
 
-@_register_op(ConvolutionForwardOpr)
+@_register_op(ConvolutionForwardOpr, ConvolutionBackwardDataOpr)
 def _convolution(opr, context):
     ph, pw = opr.ph, opr.pw
     sh, sw = opr.sh, opr.sw
     kh, kw = opr.kh, opr.kw
     param_W = opr.param_W
     group = opr.group
-    dilation_h = opr.dilation_h
+    dilation_h, dilation_w = opr.dilation_h, opr.dilation_w
+    assert (
+        dilation_h == dilation_w
+    ), "caffe accept one dilation, so dilation_h and dilation_w must equal"
     param_W = param_W.reshape((-1,) + param_W.shape[-3:])
     bias_term = opr.bias_term
     assert bias_term == False
@@ -565,9 +569,23 @@ def _convolution(opr, context):
         group=group,
         bias_term=bias_term,
     )
+    top = [context.set_blob_name(opr.out_vars[0], opr.name)]
+    if isinstance(opr, ConvolutionBackwardDataOpr):
+        layer_type = "Deconvolution"
+        bottom = [context.get_blob_name(opr.inp_vars[1])]
+    else:
+        layer_type = "Convolution"
+        bottom = [context.get_blob_name(opr.inp_vars[0])]
 
     context.add_layer(
-        _gen_layer(opr, "Convolution", context, convolution_param=param, blobs=blobs)
+        cp.LayerParameter(
+            bottom=bottom,
+            top=top,
+            name=opr.name,
+            type=layer_type,
+            blobs=blobs,
+            convolution_param=param,
+        )
     )
 
 

--- a/mgeconvert/cambricon_converter/converter.py
+++ b/mgeconvert/cambricon_converter/converter.py
@@ -311,12 +311,12 @@ def _deconv(opr, context):
         opr.name,
         inps[1],
         oups,
-        opr.stride_h,
-        opr.stride_w,
+        opr.sh,
+        opr.sw,
         opr.dilation_h,
         opr.dilation_w,
-        opr.pad_h,
-        opr.pad_w,
+        opr.ph,
+        opr.pw,
     )
     cn_deconv.param_dict["W"] = inps[0]
     b_shp = (1, inps[0].shape[0], 1, 1)

--- a/mgeconvert/mge_context/mge_op.py
+++ b/mgeconvert/mge_context/mge_op.py
@@ -120,8 +120,10 @@ class ConvolutionForwardOpr(MgeOpr):
     def __init__(self, opr):
         super().__init__(opr)
         self.kernel_shape = get_shape(opr.inputs[1])
+        self.param_W = get_symvar_value(opr.inputs[1])
         self.data_format = self.params["format"]
         self.dilation_w = self.params["dilate_w"]
+        self.dilation_h = self.params["dilate_h"]
         self.compute_mode = self.params["compute_mode"]
         self.sparse = self.params["sparse"]
 
@@ -134,12 +136,7 @@ class ConvolutionForwardOpr(MgeOpr):
             assert False, "do not support this {} format".format(self.data_format)
 
         self.num_output = get_shape(opr.outputs[0])[1]
-        self.dilation_h = self.dilation_w
-        self.param_W = get_symvar_value(opr.inputs[1])
-
         self.bias_term = False
-
-        # for caffe
         self.group = self.param_W.shape[0] if self.param_W.ndim == 5 else 1
 
 
@@ -304,17 +301,23 @@ class ConvolutionBackwardDataOpr(MgeOpr):
 
     def __init__(self, opr):
         super().__init__(opr)
-        self.kernel_shape = get_shape(opr.inputs[1])
+        # opr.inputs[0] is conv kernel
+        self.kernel_shape = get_shape(opr.inputs[0])
+        self.param_W = get_symvar_value(opr.inputs[0])
         self.data_format = self.params["format"]
         self.dilation_w = self.params["dilate_w"]
         self.dilation_h = self.params["dilate_h"]
-        self.pad_w = self.params["pad_w"]
-        self.pad_h = self.params["pad_h"]
-        self.stride_w = self.params["stride_w"]
-        self.stride_h = self.params["stride_h"]
+        self.compute_mode = self.params["compute_mode"]
         self.sparse = self.params["sparse"]
+
+        self.ph, self.pw = self.params["pad_h"], self.params["pad_w"]
+        self.sh, self.sw = self.params["stride_h"], self.params["stride_w"]
         if self.data_format == "NCHW":
             self.kh = self.kernel_shape[-2]
             self.kw = self.kernel_shape[-1]
         else:
-            raise NotImplementedError("do not support %s format" % self.data_format)
+            assert False, "do not support this {} format".format(self.data_format)
+
+        self.num_output = get_shape(opr.outputs[0])[1]
+        self.bias_term = False
+        self.group = self.param_W.shape[0] if self.param_W.ndim == 5 else 1

--- a/mgeconvert/onnx_converter/onnx_op.py
+++ b/mgeconvert/onnx_converter/onnx_op.py
@@ -13,6 +13,7 @@ from ..mge_context import (
     AxisAddRemoveOpr,
     BatchNormForwardOpr,
     ConcatOpr,
+    ConvolutionBackwardDataOpr,
     ConvolutionForwardOpr,
     DimshuffleOpr,
     ElemwiseOpr,
@@ -329,7 +330,7 @@ class DataProviderConverter(OperatorBaseConverter):
         return [], inp_tensor_list, []
 
 
-@_register_op(ConvolutionForwardOpr)
+@_register_op(ConvolutionForwardOpr, ConvolutionBackwardDataOpr)
 class Conv2DConverter(OperatorBaseConverter):
 
     __opr_type__ = "Conv"
@@ -338,7 +339,7 @@ class Conv2DConverter(OperatorBaseConverter):
         opr = self._opr
         return {
             "kernel_shape": [opr.kh, opr.kw],
-            "pads": [opr.ph, opr.ph, opr.pw, opr.pw],
+            "pads": [opr.ph, opr.pw, opr.ph, opr.pw],
             "strides": [opr.sh, opr.sw],
             "dilations": [opr.dilation_h, opr.dilation_w],
             "group": opr.group if opr.group is not None else 1,
@@ -353,21 +354,25 @@ class Conv2DConverter(OperatorBaseConverter):
         outputs = self._get_outputs()
         if attrs["group"] != 1:
             inputs[1] = opr.name + "_filter_reshape_onnx"
-            flt = opr.inp_vars[1]
+            flt = opr.param_W
             flt_shape = [
                 flt.shape[0] * flt.shape[1],
                 flt.shape[2],
                 flt.shape[3],
                 flt.shape[4],
             ]
-            flt_data = flt.np_data.reshape(flt_shape)
+            flt_data = flt.reshape(flt_shape)
             flt_tensor = onnx.helper.make_tensor_value_info(
-                inputs[1], mge2onnx_dtype_mapping[flt.dtype], flt_shape
+                inputs[1], mge2onnx_dtype_mapping[flt.dtype.type], flt_shape
             )
             flt_param = onnx.numpy_helper.from_array(flt_data, inputs[1])
             self._net_sources.append(flt_tensor)
             self._parameters.append(flt_param)
-        conv2d = onnx.helper.make_node("Conv", inputs, [outputs[0]], **attrs)
+        onnx_op = "Conv"
+        if isinstance(self._opr, ConvolutionBackwardDataOpr):
+            onnx_op = "ConvTranspose"
+            inputs = [inputs[1], inputs[0]]
+        conv2d = onnx.helper.make_node(onnx_op, inputs, [outputs[0]], **attrs)
         nodes.extend([conv2d])
         return (nodes, self._net_sources, self._parameters)
 
@@ -394,7 +399,7 @@ class Pooling2DConverter(OperatorBaseConverter):
         opr = self._opr
         attribute = {
             "kernel_shape": [opr.kh, opr.kw],
-            "pads": [opr.ph, opr.ph, opr.pw, opr.pw],
+            "pads": [opr.ph, opr.pw, opr.ph, opr.pw],
             "strides": [opr.sh, opr.sw],
         }
 

--- a/test/test_caffe.py
+++ b/test/test_caffe.py
@@ -50,11 +50,11 @@ def _test_convert_result(inputs, fpath, mge_results, max_err):
     assert np.allclose(caffe_results, mge_results, atol=max_err)
 
 
-def test_conv2d():
-    for choose in ("normal", "group"):
-        net = ConvOpr(choose)
-        mge_result = dump_mge_model(net, net.data, tmp_file)
-        _test_convert_result(net.data, tmp_file, mge_result, max_error)
+@pytest.mark.parametrize("mode", ["normal", "group", "transpose"])
+def test_conv2d(mode):
+    net = ConvOpr(mode)
+    mge_result = dump_mge_model(net, net.data, tmp_file)
+    _test_convert_result(net.data, tmp_file, mge_result, max_error)
 
 
 def test_linear():

--- a/test/test_onnx.py
+++ b/test/test_onnx.py
@@ -52,11 +52,11 @@ def _test_convert_result(inputs, fpath, mge_result, max_err):
         assert np.allclose(pred_onx, mge_result, atol=max_err)
 
 
-def test_conv2d():
-    for choose in ("normal", "group"):
-        net = ConvOpr(choose)
-        mge_result = dump_mge_model(net, net.data, tmp_file)
-        _test_convert_result(net.data, tmp_file, mge_result, max_error)
+@pytest.mark.parametrize("mode", ["normal", "group", "transpose"])
+def test_conv2d(mode):
+    net = ConvOpr(mode)
+    mge_result = dump_mge_model(net, net.data, tmp_file)
+    _test_convert_result(net.data, tmp_file, mge_result, max_error)
 
 
 def test_linear():

--- a/test/utils.py
+++ b/test/utils.py
@@ -53,8 +53,19 @@ class ConvOpr(M.Module):
         super().__init__()
         self.mode = mode
         self.data = np.random.random((1, 3, 224, 224)).astype(np.float32)
-        self.normal_conv = M.Conv2d(3, 30, 3, stride=(2, 2), padding=(2, 2))
-        self.group_conv = M.Conv2d(3, 30, 3, stride=(2, 2), padding=(2, 2), groups=3)
+        self.normal_conv = M.Conv2d(
+            3, 30, 3, stride=(2, 3), dilation=(2, 2), padding=(3, 1)
+        )
+        self.group_conv = M.Conv2d(
+            3, 30, 3, stride=(2, 3), dilation=(2, 2), padding=(3, 1), groups=3
+        )
+
+        self.transpose_conv = M.Sequential(
+            M.ConvTranspose2d(
+                3, 5, (3, 4), dilation=(2, 2), stride=(3, 2), padding=(2, 3), groups=1
+            ),
+            M.ConvTranspose2d(5, 3, (3, 3)),
+        )
 
     def forward(self, x):
         return getattr(self, self.mode + "_conv")(x)

